src/fsharp/TypeChecker.fs

```diff
@@ -5641,12 +5641,14 @@ and TcExprUndelayed cenv overallTy env tpenv (expr: SynExpr) =
         TcRecdExpr cenv overallTy env tpenv (inherits,optOrigExpr,flds,mWholeExpr)
 
     | SynExpr.While (spWhile,e1,e2,m) ->
+        let env = ClearContext env
```

Comment:
it's not easy. we have to be very careful to not erase from the wrong position.
Actually I'm not even convinced if the "context" is the best idea here.